sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -508,26 +518,30 @@ private static long getNumSplits(
 
     @Override
     public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
-      try (Cluster cluster =
-          getCluster(
-              spec.hosts(),
-              spec.port(),
-              spec.username(),
-              spec.password(),
-              spec.localDc(),
-              spec.consistencyLevel())) {
-        if (isMurmur3Partitioner(cluster)) {
-          try {
-            List<TokenRange> tokenRanges =
-                getTokenRanges(cluster, spec.keyspace().get(), spec.table().get());
-            return getEstimatedSizeBytesFromTokenRanges(tokenRanges);
-          } catch (Exception e) {
-            LOG.warn("Can't estimate the size", e);
+      if (estimatedSize != null) {
+        return estimatedSize;
+      } else {
+        try (Cluster cluster =
+            getCluster(
+                spec.hosts(),
+                spec.port(),
+                spec.username(),
+                spec.password(),
+                spec.localDc(),
+                spec.consistencyLevel())) {
+          if (isMurmur3Partitioner(cluster)) {
+            try {
+              List<TokenRange> tokenRanges =
+                  getTokenRanges(cluster, spec.keyspace().get(), spec.table().get());
+              return getEstimatedSizeBytesFromTokenRanges(tokenRanges);
```

Comment:
Can we also cache the size here in case `getEstimateSizeBytes` is called more than once?

Reply:
done